Staging SFT training

rakkit opened the pull request:
### TL;DR
Adds SFT training to Torchtitan plus a small `greedy_packing` addition. 

## **Most of code borrowed from  [Verl](https://github.com/volcengine/verl/blob/main/verl/utils/dataset/multiturn_sft_dataset.py) and [OpenRLHF](https://github.com/OpenRLHF/OpenRLHF/blob/main/openrlhf/datasets/sft_dataset.py#L35)**

### Changes
- Added SFT dataset config to job config
- Updated attention + modifying `get_attention_masks` to support SFT masks (only landed on Llama3 now)
- Temporarily use `HFTokenizer`,  need to fix this later
- Added SFT dataset/dataloader that returns `input_ids`, `labels` (user tokens masked), `attention_masks`, `position_ids`

###  TODO
- Need to verify PP / CP
- Val dataset not working yet (may depend on https://github.com/pytorch/torchtitan/pull/2142)

### Run
```bash
CONFIG_FILE="./torchtitan/models/llama3/train_configs/debug_model.toml" NGPU=4 ./run_train.sh \
  --training.running_sft_training \
  --model.flavor=debugmodel_varlen_attn \
  --training.dataset_path=openai/gsm8k \
  --sft_data_config.dataset_subset=main
  ```
  
<img width="2382" height="322" alt="image" src="https://github.com/user-attachments/assets/a2e94c12-4358-48eb-a11e-14161f9f52c2" />


## more test
```
CONFIG_FILE="./torchtitan/models/llama3/train_configs/llama3_8b.toml" NGPU=4 ./run_train.sh \
 --training.running_sft_training \ --model.flavor=8B_varlen \ --training.dataset_path=openai/gsm8k \
 --sft_data_config.dataset_subset=main \
 --model.hf_assets_path="$Home/Meta-Llama-3.1-8B-Instruct/" \
 --training.local_batch_size=8 \ --activation_checkpoint.mode=full \
 --training.seq_len=2048 \ --training.steps=100 \
 --lr_scheduler.warmup_steps=10 \  --debug.seed 10 \
 --sft_data_config.pad_mode=right_padding \
 --metrics.enable_wandb \
 ```
 (`torch                    2.10.0.dev20251124+cu129` and i am using cudnn attention )
<img width="2470" height="782" alt="image" src="https://github.com/user-attachments/assets/aae96d77-8c68-4fd1-a6af-9da5461046bb" />

<img width="5056" height="2656" alt="W B Chart 12_12_2025, 8_32_32 PM" src="https://github.com/user-attachments/assets/587688d0-1bce-43d7-a65c-e011935746f2" />

compile does not work for no-padding because the seq-len for each training step keeps changing. We could pad the buffer to `seqlen` when turning on `greedy_packing`. (its `packing_on++`) to make compile happy. 

<img width="2478" height="812" alt="image" src="https://github.com/user-attachments/assets/8bd1b716-3270-47a4-bb8c-fa196ed91bd1" />

rakkit committed:
add SFT dataset, configs and adapt tokenizer/model to work with sft

tianyu-l commented on torchtitan/hf_datasets/sft_text_datasets.py:
```diff
@@ -0,0 +1,559 @@
+    return system_prompt, generate_prompt
+
+
+def preprocess_data(
```
Is this function used?

rakkit replied:
it used in SFTdataset [init](https://github.com/rakkit/torchtitan/blob/staging_sft/torchtitan/hf_datasets/sft_text_datasets.py#L156)  to infer the "system_prompt" and "generation_prompt", the length of these two prompts will be used to remove the repeated [SYSTEM] prompt in multiturn data.
e.g.  if will do `tokenzier(message_i) for message_i in row` we will get
`[SYSTEM …][USER …][SYSTEM …][ASSISTANT …][SYSTEM …][USER …]...`

So we want to remove the redundant `[SYSTEM …]` since the second message.
(condition on `index` of[ `_process_single_message` func](
https://github.com/rakkit/torchtitan/blob/staging_sft/torchtitan/hf_datasets/sft_text_datasets.py#L237))

tianyu-l commented on torchtitan/components/tokenizer.py:
```diff
@@ -409,6 +432,10 @@ def id_to_token(self, token_id: int) -> Optional[str]:
 
+    def apply_chat_template(self, conversation, **kwargs):
+
+        return self.backup_hf_tokenizer.apply_chat_template(conversation, **kwargs)
```
I learned what this doing from https://huggingface.co/docs/transformers/en/chat_templating

> (Be aware that when it turns on apply_chat_template, we are supposed to provide the chat_template.jinja in tokenzier's folder. There is no such file for "Meta-Llama-3.1-8B-Instruct". For test purposes, you can use, e.g,. tokenizer from [Olmo-3-7B-Instruct](https://huggingface.co/allenai/Olmo-3-7B-Instruct)

This sounds confusing. If it's supposed to function, why it's missing a file?

wwwjn replied:
I think this `backup_hf_tokenizer` will fragment and causing potential issue? We use torchtitan tokenizer for encode/decode, but relay on `backup_hf_tokenizer` to apply_chat_template

tianyu-l commented on torchtitan/hf_datasets/sft_text_datasets.py:
```diff
@@ -0,0 +1,559 @@
+from torchtitan.config.job_config import SFTDataConfig
+from torchtitan.tools.logging import logger
+
+IGNORE_INDEX = -100
```
This is also defined at https://github.com/pytorch/torchtitan/blob/c1f4e9491e224acadd69b8e59f8abd9b0aae7145/torchtitan/experiments/vlm/infra/loss.py#L21
Shall we put into a more central place?

tianyu-l commented on torchtitan/components/tokenizer.py:
```diff
@@ -59,6 +60,8 @@ def __init__(
         self.bos_token = None
         self.eos_token = None
+        self.pad_id = None  # only used for SFT
+        self.pad_token = None  # only used for SFT
```
Let's have a separate (sub)class for SFT tokenizer, and not depending on `transformers`.

See related comments https://github.com/pytorch/torchtitan/pull/1802#discussion_r2423423920

tianyu-l commented on torchtitan/models/attention.py:
```diff
@@ -272,6 +272,28 @@ def sliding_window_mod(
     return sliding_window_mod
 
 
+def get_sft_padding_mask_mod(attention_masks: torch.Tensor) -> _mask_mod_signature:
```
I don't think we want to materialize an attention_masks to general mask_mod. It defeats the purpose of why using mask_mod (to represent mask sparsely).

Instead, we should use e.g. pad_id to infer the mask_mod here, similar to how we generate document mask using eos_id.

Similar for varlen metadata generation.

rakkit replied:
you mean `<|eos|>` or `eos_token`?  
`eos_token` will not work because some SFT/PT model `eos_token` does not means end of "documents"

`<|eos|>` can work, but we need to ask the user to explicitly provide a segment token, and we need to wrap around tokenizer and `get_document_mask_mod` to get the correct "segment token."


> I actually never seen this practice because usually people create another token like <|end_of_turn|> or <|im_end|> or something, but I might be wrong.

I think that's the argument, if we assume users are well known on SFT so can set it properly, or we wanna make sure the training can be done correctly, even for users who dont know anything.  (I am ok for both)

tianyu-l commented on torchtitan/config/job_config.py:
```diff
@@ -198,11 +198,36 @@ class LRScheduler:
 
 
+@dataclass
+class SFTDataConfig:
```
Due the limitation of current config system, we can't express that "we should either use a (default) Data when doing pretraining, or use the SFTData config when doing finetuning". Right now you'd have to register the SFT one even when doing pretraining.

We are thinking of redo the config system and leaning towards "python as the config". cc @shuhuayu

tianyu-l commented on torchtitan/hf_datasets/text_datasets.py:
```diff
@@ -178,21 +178,42 @@ def build_text_dataloader(
+            dp_world_size=dp_world_size,
+            infinite=infinite,
+        )
+    else:
```
Let's create another `build_dataloader_fn` (so that one can choose during TrainSpec creation) instead of merging everything together with if-else. And yes, it means we may need to create another experiment folder, which I don't mind we start from. That way you could also define the SFTDataConfig in the folder.

Eventually we should revamp torchtitan config system. cc @shuhuayu

tianyu-l reviewed (commented):
Thanks, left some initial comments.

tianyu-l commented on torchtitan/models/attention.py:
```diff
@@ -340,3 +362,70 @@ def create_varlen_metadata_for_document(
     )
+
+
+def create_varlen_metadata_for_sft(
```
from what I heard, right now varlen may be using flash v2 so known to be slower than flex. Flash v4 and cudnn backend are on the way (and should be faster)

rakkit commented on torchtitan/hf_datasets/sft_text_datasets.py:
```diff
@@ -0,0 +1,559 @@
+
+        if self.pad_mode == "no_padding":
+            self.collate_fn = partial(
+                collate_sft_batch, pad_token_id=self.tokenizer.pad_id
```
Sounds reasonable.   For whatever reason, people need opt-1; they should just try right-padding.

lkhphuc commented:
> should we filter out ids!= IGNORE_INDEX to get the "correct" ntokens_seen?

No IMO. It's "token seens", not "token computed loss on" afterall. We just need to ensure  padding is removed from token seen.

wwwjn commented on torchtitan/models/test_mask.py:
Should we move this file under `tests/unit_tests`?

wwwjn commented on torchtitan/models/attention.py:
```diff
@@ -215,6 +218,92 @@ def document_mask(
+        assert seq_lens is not None and seq_start_indices is not None
+        return get_document_mask_mod_using_sequence_indices(
+            batch, seq_lens, seq_start_indices
+        )
```
Error-out if both `eos_id` and `extra_inputs` are empty?

wwwjn reviewed (commented):
Thanks for the work! Structurally, I would advocate if we start in an experiment folder, so we could highly increase readability for users. We could have something like 
- `experiement/sft`:
    - `tokenizer.py`: Inherit from torchtitan's `HuggingFaceToenizer` , or depend on AutoTokenizer 
    - `datasets.py`
    - `job_config.py`
    - `__init__.py `: use TrainSpec to plug in dataloader, tokenizer here

Currently branching in main trainer / job_config would seems confusing to me

rakkit commented on torchtitan/hf_datasets/sft_text_datasets.py:
```diff
@@ -0,0 +1,562 @@
+
+    def _process_one_row(self, row_dict: dict):
+        """Convert a dataset row into model-ready tensors with causal labels."""
+        messages = _build_messages(
```
yes thats smart. thx

rakkit commented:
Thanks again for the feedback ❤️.  Sft code now moved to `experiments` folder, also simplified lots of configs.
I will try to edit and improve documents. 

To run test on GSM8K:
```
CONFIG_FILE="./torchtitan/models/llama3/train_configs/debug_model.toml" NGPU=4 ./run_train.sh \
 --model.flavor=debugmodel_varlen_attn \
 --training.dataset question_answer \
 --training.dataset_path=openai/gsm8k \
 --sft_config.dataset_subset=main \
 --model.name=sft.llama3 --job.custom_config_module=torchtitan.experiments.sft.job_config \
```

and to run test on `Dolci-Instruct-SFT` with chat template, (llama3-8b-flex)
```

rm -rf ./outputs/*
CONFIG_FILE="./torchtitan/models/llama3/train_configs/llama3_8b.toml" NGPU=4 ./run_train.sh \
 --model.name=sft.llama3 --job.custom_config_module=torchtitan.experiments.sft.job_config \
 --model.flavor=8B_flex --training.dataset=multi_turn --sft_config.apply_chat_template \
 --training.dataset_path=$HOME/Dolci-Instruct-SFT/ \
  --model.hf_assets_path=$HOME//Meta-Llama-3.1-8B-Instruct/ \
 --training.local_batch_size=8 \
 --activation_checkpoint.mode=full \
 --metrics.log_freq=10 \
 --training.seq_len=2048 --training.steps=100 --lr_scheduler.warmup_steps=10 \
 --debug.seed 10 \
 --compile.enable \

```

acisseJZhong commented on torchtitan/experiments/sft/sft_text_datasets.py:
```diff
@@ -0,0 +1,640 @@
+        - Either tokenize the message["content"] directly, OR
+        - Apply the tokenizer's chat template (requires the chat template to be configured).
+
+   3.2) If there are MULTIPLE messages, we call `_process_single_message` for each message.
```
wondering instead of removing system prompt later, would it be possible to add a flag to _process_single_message to control if we want to add the system prompt?

tianyu-l commented on torchtitan/experiments/sft/job_config.py:
```diff
@@ -0,0 +1,71 @@
+    """Subset to use"""
+    stream_dataset: bool = True
+    """Whether to stream the dataset"""
+    is_multiturn: bool = False
```
is this field used?

tianyu-l commented on torchtitan/experiments/sft/sft_text_datasets.py:
```diff
@@ -0,0 +1,640 @@
+        streaming=sft_config.stream_dataset,
+    )
+
+    message_builder = DATASET_MESSAGE_BUILDERS[dataset_name]
```
the key should be coming from a SFTConfig e.g. `task_type`, not using `dataset_name` which was originally used for dataset registration https://github.com/pytorch/torchtitan/blob/main/torchtitan/hf_datasets/text_datasets.py#L36

tianyu-l commented on torchtitan/experiments/sft/sft_text_datasets.py:
```diff
@@ -0,0 +1,640 @@
+        _build_prompt_response_messages_from_row_dict,
+        prompt_key="question",
+        response_key="answer",
+    ),
```
This separation sounds dataset specific. Without specifying dataset it's hard to tell if they are general.

Maybe consider registering a DATASETS field similar to https://github.com/pytorch/torchtitan/blob/main/torchtitan/hf_datasets/text_datasets.py#L36
if you have good candidate

tianyu-l commented on torchtitan/experiments/sft/sft_text_datasets.py:
```diff
@@ -0,0 +1,640 @@
+
+
+DATASET_MESSAGE_BUILDERS = {
+    "multi_turn": _build_multi_turn_messages_from_row_dict,
```
similar to this one -- too abstract without seeing an example and hard to verify if it's correct

tianyu-l commented on torchtitan/experiments/sft/llama3/__init__.py:
```diff
@@ -0,0 +1,37 @@
+        build_lr_schedulers_fn=build_lr_schedulers,
+        build_dataloader_fn=build_sft_text_dataloader,
+        build_tokenizer_fn=build_auto_tokenizer,
+        build_loss_fn=build_cross_entropy_loss,
```
You may want to use https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/vlm/infra/loss.py#L100

otherwise with DP / CP it won't be correct.

acisseJZhong commented on torchtitan/experiments/sft/auto_tokenizer.py:
```diff
@@ -0,0 +1,90 @@
+    def decode(self, tokens: list[int], *args, **kwargs):
+        decoded = self.tokenizer.decode(tokens, *args, **kwargs)
+        # this is an ad-hoc for debugging purpose
+        if self.pad_id in tokens:
```
can you add more comment about what is this for?

rakkit commented on torchtitan/experiments/sft/auto_tokenizer.py:
```diff
@@ -0,0 +1,90 @@
+        pad_token: str | None = None,
+    ):
+        self.tokenizer = HF_AutoTokenizer.from_pretrained(
+            tokenizer_path, eos_token=eos_token, use_fast=True
```
you might be right.   we need to know "eos_token" and "eos_id" but not necessarily override HF's tokenizer. i will check the code logic again

acisseJZhong commented on torchtitan/experiments/sft/sft_text_datasets.py:
```diff
@@ -0,0 +1,622 @@
+                    self._sample_idx += 1
+                    continue
+
+                if self._buffer["current_len"] + new_len > self.buffer_max_length:
```
is this greedy packing? maybe put it inside else?

rakkit commented on torchtitan/experiments/sft/auto_tokenizer.py:
```diff
@@ -0,0 +1,90 @@
+        self.bos_token = self.tokenizer.bos_token
+        self.eos_token = self.tokenizer.eos_token
+
+        self.pad_id = pad_token_id
```
yes and it gonna bit of messy.  In short we should try to document everything and let user know what they are doing and we assume user should have some knowledge and can understand what will happen in this SFT receipt after read the document. (i can write documents).

For `pad`, we should "assume" it is not used in pre-training or even 'sft'.  it suppose to be a "meaningless" token and only used for packing.  And we completely drops pad after sft-training.

For `eos`, this is really complicated and really depends on chat-template.

In general and common setup, we can consider to have conservation like <|start|>assistant xxxxxx <|end|>,  that <|end|> is end of sentence (count in loss).   And in this case, we append a "EOS" that only used for indicating split of document, to make it like:

[system] [user-1] [assisant-1] [user-2] <|start|>assistant xxxxxx <|end|>  "EOS"

where  is not count in loss.   
The option of "EOS" can be either the EOS in pre-training, or any of reserved token that have no meaning, but it cannot be <|end|> and cannot be token can appear in middle of sequence.     The <|end|> token should be configured in chat-template and we dont have to bother. After SFT,  user should set <|end|> as the actual "stop" token in generation, and optionally to have  (which should not affect).

There is always special cases (not sure we want to support them in first place), especially, gpt-oss and harmony format.  They explicitly want  as stop token, that above example should be
[system] [user-1] [assisant-1] [user-2] <|start|>assistant xxxxxx "EOS"    and here  "EOS" should be count in loss.

rakkit commented on torchtitan/experiments/sft/job_config.py:
```diff
@@ -0,0 +1,80 @@
+    Whether to ignore the input_ids mismatch when apply chat template.
+    """
+
+    eos_token: str | None = None
```
you are right, i forget to fix job_configs.  

we should explicitly ask user to set "eos_token" and "pad_token_id"

acisseJZhong commented on torchtitan/experiments/sft/sft_text_datasets.py:
```diff
@@ -0,0 +1,622 @@
+                [message],
+                tools=tools,
+                add_generation_prompt=False,
+                tokenize=True,
```
instead of removing system prompt later, can you set this to False except for the first call? cc @tianyu-l

rakkit replied:
you mean if there is an flag in `apply_chat_template` to not add system prompt?  
For that answer is kind "no", the logic of what to do with "message" is mainly coded in the actual chat-template not this `apply_chat_template` functions.   Unless some chat-tempalte have this logic to allow "disable" system prompt.


it seems as long as we process message one by one it is unavoidable to remove system prompt later.

acisseJZhong replied:
ic, makes sense. I somehow misread this `tokenizer` keyword to be whether if it will append system prompt or not.

tianyu-l commented on torchtitan/experiments/sft/sft_text_datasets.py:
```diff
@@ -0,0 +1,622 @@
+        There is one special case:
+        When `apply_chat_template=True`, each call to `_process_single_message` may add a
+        [system prompt]. That can produce:
+        [System] [User] [System] [Assistant] [System] [User] [System] [Assistant] ...
```
> For for whatever reason, in verl where i followed their conversation. They decide to process the dict in list one by one.

So it looks like that we followed verl impl which process the list elements one-by-one, so that we have to remove excess "system".

Are their alternatives than verl -- what's the down side of doing 

> So HF's tokenizer can actually read and render this entire lists of dictionary in one place and gives
[System] - {user} - {assistant} - {user} - {assistant}. This is exactly what we want.

Sorry if this has been answered elsewhere.

garrett361 commented:
> and you might be interested in @garrett361

Thanks for the ping. And yep, just took a quick glance, but this all looks very familiar haha. (My SFT-in-titan impls were less mature I'm sure.)

There's a lot here. Is there anything you want particular comment on @rakkit ?

I know one thing that can cause many headaches is the loss reduction. If you want the same training dynamics across different training setups, you need to be very careful about how loss averages are computed. Very easy to implement this incorrectly so that your loss (assuming you want a mean-loss and not a sum-loss like OLMO uses, say) is not actually the mean loss over the world job, and is instead some incorrect mean-of-means.  

You might have covered this already. Didn't go through in detail.

tianyu-l commented:
SFT should be supported now so closing

vadimkantorov commented:
@tianyu-l Could you please advise, is there anywhere guide/docs for the new SFT support? Thanks!

tianyu-l commented:
Here's an example https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/llama3/config_registry.py#L349

vadimkantorov commented:
It does support CE masks, right? (to skip training on the prompt prefix)
